### PR TITLE
Activity for HttpClient calls not stopped if inner handler does not use async/await (PR 2)

### PR DESCRIPTION
**Pull request by cwe1ss**
Note: This is a new version of #27472. I restored my branch but the PR didn't pick up the new commits. Sorry for that.

Original description from #27472:

I've been writing some `DiagnosticSource`/`Activity` based instrumentation for HttpClient. [In my unit tests](https://github.com/cwe1ss/opentracing-contrib-dotnet/blob/ba20de4ba57e43cb685083e3c0b2b19022ce067e/test/OpenTracing.Contrib.AspNetCore.Tests/HttpOut/HttpOutInterceptorTest.cs) I've created a HttpMessageHandler and wrapped it with the existing `DiagnosticsHandler` (by using reflection, so in an unsupported way) to test the different scenarios.

However, I noticed that if my handler doesn't use `async/await`, the `DiagnosticsHandler` does NOT raise the Diagnostics-events for `System.Net.Http.Exception` and `System.Net.Http.HttpRequestOut.Stop` because the invocation is outside of the try...catch block.

E.g. if DiagnosticsHandler is using an inner handler that looks like this, it does NOT raise the events:
```csharp
public class FailingHttpMessageHandler : HttpMessageHandler
{
    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        throw new InvalidOperationException("foo");
    }
}
```

It only works, if I change the code as follows:
```csharp
public class FailingHttpMessageHandler : HttpMessageHandler
{
    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        await Task.CompletedTask;
        throw new InvalidOperationException("foo");
    }
}
```

My direct usage of DiagnosticsHandler is obviously a hack as it is internal and AFAIU in a regular application `DiagnosticsHandler` is an internal feature of `HttpClientHandler` so there shouldn't be any user code after it but nevertheless, I think that it would make sense to move the call from this PR into the try..catch block.

/cc @lmolkova @vancem

### CODE REVIEW

**Comment by cwe1ss** [comment]
The new commits now compile on my machine and I've added a test that demonstrates the problem by producing an invalid state for the inner handler. If you run the test on the current code base without the fix, it would fail.

I had to write different logic for SocketsHttpHandler and WinHttpHandler because they seem to have different behavior.

**Comment by vancem** [comment]
It looks good to me too.

**Comment by vancem** [comment]
Seems we had an infrastructure race issue?    Trying again to see if it is deterministic. 

@dotnet-bot  test NETFX x86 Release Build

**Comment by cwe1ss** [comment]
Is there anything I should/can do here? Do you think this is something related to my change? Even from `master` I sometimes get timeouts locally as well if I run `build.cmd src/System.Net.Http -OuterLoop` - e.g.:

```
  call C:\src\github\cwe1ss\corefx\bin\testhost\netcoreapp-Windows_NT-Debug-x64\\dotnet.exe xunit.console.netcore.exe System.Net.Http.Functional.Tests.dll  -xml testResults.xml -notrait category=nonnetcoreapptests -notrait category=nonw
  indowstests  -notrait category=failing
  popd
  ===========================================================================================================
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.

  Discovering: System.Net.Http.Functional.Tests
  Discovered:  System.Net.Http.Functional.Tests
  Starting:    System.Net.Http.Functional.Tests
     System.Net.Http.Functional.Tests.DefaultCredentialsTest.Proxy_UseAuthenticatedProxyWithWrappedDefaultCredentials_OK [SKIP]
        Condition(s) not met: \"DomainProxyTestsEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientEKUTest.HttpClient_ServerEKUClientAuth_Fails [SKIP]
        Condition(s) not met: \"CanTestClientCertificates\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientEKUTest.HttpClient_NoEKUServerAuth_Ok [SKIP]
        Condition(s) not met: \"CanTestCertificates\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientEKUTest.HttpClient_NoEKUClientAuth_Ok [SKIP]
        Condition(s) not met: \"CanTestClientCertificates\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientEKUTest.HttpClient_ClientEKUServerAuth_Fails [SKIP]
        Condition(s) not met: \"CanTestCertificates\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.ManyClients_ManyPosts_Async [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.CreateAndDestroyManyClients [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.ManyClients_ManyGets [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.MakeAndFaultManyRequests [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.SingleClient_ManyGets_Async [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.SingleClient_ManyGets_Sync [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientMiniStress.UnreadResponseMessage_Collectible [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.ManyClients_ManyGets [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.CreateAndDestroyManyClients [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.SingleClient_ManyGets_Async [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.UnreadResponseMessage_Collectible [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.MakeAndFaultManyRequests [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.SingleClient_ManyGets_Sync [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientMiniStress.ManyClients_ManyPosts_Async [SKIP]
        Condition(s) not met: \"IsStressModeEnabled\"
     System.Net.Http.Functional.Tests.HttpClientEKUTest.HttpClient_ClientEKUServerAuth_Fails [SKIP]
        Condition(s) not met: \"CanTestCertificates\"
     System.Net.Http.Functional.Tests.HttpClientEKUTest.HttpClient_NoEKUClientAuth_Ok [SKIP]
        Condition(s) not met: \"CanTestClientCertificates\"
     System.Net.Http.Functional.Tests.HttpClientEKUTest.HttpClient_NoEKUServerAuth_Ok [SKIP]
        Condition(s) not met: \"CanTestCertificates\"
     System.Net.Http.Functional.Tests.HttpClientEKUTest.HttpClient_ServerEKUClientAuth_Fails [SKIP]
        Condition(s) not met: \"CanTestClientCertificates\"
     System.Net.Http.Functional.Tests.SocketsHttpHandler_DefaultCredentialsTest.Proxy_UseAuthenticatedProxyWithWrappedDefaultCredentials_OK [SKIP]
        Condition(s) not met: \"DomainProxyTestsEnabled\"
     System.Net.Http.Functional.Tests.HttpClientHandlerTest.Proxy_BypassFalse_GetRequestGoesThroughCustomProxy_NonRfcCompliant(creds: null, wrapCredsInCache: False) [FAIL]
        System.Net.Http.HttpRequestException : An error occurred while sending the request.
        ---- System.Net.Http.WinHttpException : The operation timed out
        Stack Trace:
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs(469,0): at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean
   disposeCts)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\tests\FunctionalTests\LoopbackGetRequestHttpProxy.cs(142,0): at System.Net.Http.Functional.Tests.LoopbackGetRequestHttpProxy.StartAsync(TcpListener listener, Boolean requireAuth
  , Boolean expectCreds)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(2842,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.Proxy_BypassFalse_GetRequestGoesThroughCustomProxy_Implementati
  on(ICredentials creds, Boolean wrapCredsInCache)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(2814,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.Proxy_BypassFalse_GetRequestGoesThroughCustomProxy_NonRfcCompli
  ant(ICredentials creds, Boolean wrapCredsInCache)
           --- End of stack trace from previous location where exception was thrown ---
           ----- Inner Stack Trace -----
           C:\src\github\cwe1ss\corefx\src\System.Runtime.Extensions\src\System\Environment.cs(146,0): at System.Environment.get_StackTrace()
           C:\src\github\cwe1ss\corefx\src\Common\src\System\Runtime\ExceptionServices\ExceptionStackTrace.cs(23,0): at System.Runtime.ExceptionServices.ExceptionStackTrace.AddCurrentStack(Exception exception)
           C:\src\github\cwe1ss\corefx\src\Common\src\System\Net\Http\WinHttpException.cs(56,0): at System.Net.Http.WinHttpException.CreateExceptionUsingError(Int32 error)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs(324,0): at System.Net.Http.WinHttpRequestCallback.OnRequestError(WinHttpRequestState state, WINHTTP_ASYNC_RESULT asy
  ncResult)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs(104,0): at System.Net.Http.WinHttpRequestCallback.RequestCallback(IntPtr handle, WinHttpRequestState state, UInt32 i
  nternetStatus, IntPtr statusInformation, UInt32 statusInformationLength)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs(47,0): at System.Net.Http.WinHttpRequestCallback.WinHttpCallback(IntPtr handle, IntPtr context, UInt32 internetStatu
  s, IntPtr statusInformation, UInt32 statusInformationLength)
           --- End of stack trace from AddCurrentStack ---
           C:\src\github\cwe1ss\corefx\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs(62,0): at System.Threading.Tasks.RendezvousAwaitable`1.GetResult()
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(856,0): at System.Net.Http.WinHttpHandler.StartRequest(WinHttpRequestState state)
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandlerTest.GetAsync_RequestHeadersAddCustomHeaders_HeaderAndValueSent(name: \"X-CustomHeader\", value: \"x-value\", uri: http://corefx-net.cloudapp.net/Redirect.ashx?st
  atuscode=302&uri=http%3A%2F%2Fcorefx-net.cloudapp.net%2FEcho.ashx) [FAIL]
        System.Net.Http.HttpRequestException : A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
        ---- System.Net.Sockets.SocketException : A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
        Stack Trace:
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\ConnectHelper.cs(98,0): at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
           E:\A\_work\1819\s\src\mscorlib\shared\System\Threading\Tasks\ValueTask.cs(713,0): at System.Threading.Tasks.ValueTask`1.get_Result()
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs(309,0): at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken ca
  ncellationToken)
           E:\A\_work\1819\s\src\mscorlib\shared\System\Threading\Tasks\ValueTask.cs(713,0): at System.Threading.Tasks.ValueTask`1.get_Result()
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs(450,0): at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
           E:\A\_work\1819\s\src\mscorlib\shared\System\Threading\Tasks\ValueTask.cs(713,0): at System.Threading.Tasks.ValueTask`1.get_Result()
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs(243,0): at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth,
  CancellationToken cancellationToken)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\RedirectHandler.cs(30,0): at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs(469,0): at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean
   disposeCts)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(1131,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.GetAsync_RequestHeadersAddCustomHeaders_HeaderAndValueSent(Stri
  ng name, String value, Uri uri)
           --- End of stack trace from previous location where exception was thrown ---
           ----- Inner Stack Trace -----
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\ConnectHelper.cs(78,0): at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
     System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandlerTest.Ctor_ExpectedDefaultPropertyValues_UapPlatform [SKIP]
        Condition(s) not met: \"IsUap\"
     System.Net.Http.Functional.Tests.HttpClientHandlerTest.PostAsync_CallMethodTwice_StringContent(remoteServer: https://corefx-net.cloudapp.net/VerifyUpload.ashx) [FAIL]
        System.Net.Http.HttpRequestException : An error occurred while sending the request.
        ---- System.Net.Http.WinHttpException : The operation timed out
        Stack Trace:
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs(469,0): at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean
   disposeCts)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http\tests\FunctionalTests\HttpClientHandlerTest.cs(2082,0): at System.Net.Http.Functional.Tests.HttpClientHandlerTest.PostAsync_CallMethodTwice_StringContent(Uri remoteServer)
           --- End of stack trace from previous location where exception was thrown ---
           ----- Inner Stack Trace -----
           C:\src\github\cwe1ss\corefx\src\System.Runtime.Extensions\src\System\Environment.cs(146,0): at System.Environment.get_StackTrace()
           C:\src\github\cwe1ss\corefx\src\Common\src\System\Runtime\ExceptionServices\ExceptionStackTrace.cs(23,0): at System.Runtime.ExceptionServices.ExceptionStackTrace.AddCurrentStack(Exception exception)
           C:\src\github\cwe1ss\corefx\src\Common\src\System\Net\Http\WinHttpException.cs(56,0): at System.Net.Http.WinHttpException.CreateExceptionUsingError(Int32 error)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs(324,0): at System.Net.Http.WinHttpRequestCallback.OnRequestError(WinHttpRequestState state, WINHTTP_ASYNC_RESULT asy
  ncResult)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs(104,0): at System.Net.Http.WinHttpRequestCallback.RequestCallback(IntPtr handle, WinHttpRequestState state, UInt32 i
  nternetStatus, IntPtr statusInformation, UInt32 statusInformationLength)
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs(47,0): at System.Net.Http.WinHttpRequestCallback.WinHttpCallback(IntPtr handle, IntPtr context, UInt32 internetStatu
  s, IntPtr statusInformation, UInt32 statusInformationLength)
           --- End of stack trace from AddCurrentStack ---
           C:\src\github\cwe1ss\corefx\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs(62,0): at System.Threading.Tasks.RendezvousAwaitable`1.GetResult()
           C:\src\github\cwe1ss\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs(856,0): at System.Net.Http.WinHttpHandler.StartRequest(WinHttpRequestState state)
     System.Net.Http.Functional.Tests.HttpClientHandlerTest.Ctor_ExpectedDefaultPropertyValues_UapPlatform [SKIP]
        Condition(s) not met: \"IsUap\"
  Finished:    System.Net.Http.Functional.Tests

  === TEST EXECUTION SUMMARY ===
C:\src\github\cwe1ss\corefx\Tools\tests.targets(492,5): warning :    System.Net.Http.Functional.Tests  Total: 1968, Errors: 0, Failed: 3, Skipped: 26, Time: 297,680s [C:\src\github\cwe1ss\corefx\src\System.Net.Http\tests\FunctionalTests
\System.Net.Http.Functional.Tests.csproj]
```

**Comment by vancem** [comment]
I am OK with this change.   @davidsh  do you want anyone on your team to review, or can we pull the trigger?

**Comment by davidsh** [comment]
>I am OK with this change. @davidsh do you want anyone on your team to review, or can we pull the trigger?

I think it would be useful if @stephentoub looked at this especially since the PR is trying to solve some issues with async await patterns.

File: src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs [hunk 1]
```diff
@@ -92,10 +92,14 @@ internal static bool IsEnabled()
...
 
-            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+            Task<HttpResponseMessage> responseTask = null;
             try
             {
+                responseTask = base.SendAsync(request, cancellationToken);
+
                 await responseTask.ConfigureAwait(false);
+
+                return responseTask.Result;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs]: This can just be:
```C#
return await responseTask.ConfigureAwait(false);
```

**Comment by stephentoub** [comment]
> However, I noticed that if my handler doesn't use async/await, the DiagnosticsHandler does NOT raise the Diagnostics-events for System.Net.Http.Exception and System.Net.Http.HttpRequestOut.Stop because the invocation is outside of the try...catch block.

To be clear, it's not really about async/await, rather it's about whether exceptions can synchronously emerge from the SendAsync method.  When using `async`, it effectively surrounds the whole method with a try/catch and returns any thrown exception inside of the returned Task, rather than allowing them to propagate.  The same thing can be done without async/await.

File: src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs (hunk 1 above)
Comment (cwe1ss) [reply src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs]: Thank you for the feedback, I fixed this.

**Comment by cwe1ss** [comment]
@dotnet-bot help

**Comment by cwe1ss** [comment]
Awesome, thank you!!!

**Pull request merged**